rpcs3/Emu/RSX/overlays.cpp

```diff
@@ -20,12 +27,215 @@ namespace rsx
+			m_titles.back_color     = {0.0f, 0.0f, 0.0f, 0.0f};
+			m_titles.padding_top    = 0;
+			m_titles.padding_bottom = -14;
+			m_titles.set_margin(5, 5, 0, 0);
+
+			switch (m_detail)
+			{
+			case detail_level::minimal_1:
+			case detail_level::low_2: m_titles.text = ""; break;
+			case detail_level::medium_3: m_titles.text = fmt::format("\n\n%s", title1_medium); break;
+			case detail_level::high_4: m_titles.text = fmt::format("\n\n%s", title1_high); break;
+			case detail_level::ultra_5: m_titles.text = fmt::format("\n\n%s\n\n\n\n\n%s", title1_high, title2); break;
+			}
+			m_titles.auto_resize();
+			m_titles.refresh();
+		}
+
+		void perf_metrics_overlay::init()
+		{
+			reset_titles();
+			reset_body();
+
+			update(true);
+			m_update_timer.Start();
+		}
+
+		perf_metrics_overlay::perf_metrics_overlay(bool initialize)
+		{
+			// Default values, will change based on config options
+			m_update_interval = 350;
+			m_detail          = detail_level::ultra_5;
+			m_font_size       = 11;
+
+			if (initialize)
+				init();
+		}
+
+		// In ms
+		void perf_metrics_overlay::set_update_interval(u32 update_interval)
+		{
+			m_update_interval = update_interval;
+		}
+
+		void perf_metrics_overlay::set_detail_level(detail_level level)
+		{
+			m_detail = level;
+			reset_titles();
+		}
+
+		void perf_metrics_overlay::set_font_size(u32 font_size)
+		{
+			m_font_size = font_size;
+			reset_titles();
+			reset_body();
+		}
+
+		void perf_metrics_overlay::update(bool force_update)
+		{
+			const auto elapsed = m_update_timer.GetElapsedTimeInMilliSec();
+
+			if (!force_update)
+			{
+				++m_frames;
+			}
+
+			if (elapsed >= m_update_interval || force_update)
+			{
+				f32 fps{0};
+				f32 frametime{0};
+
+				u64 ppu_cycles{0};
+				u64 spu_cycles{0};
+				u64 rsx_cycles{0};
+				u64 total_cycles{0};
+
+				u32 ppus{0};
+				u32 spus{0};
+				u32 rawspus{0};
+
+				f32 cpu_usage{-1.f};
+				u32 total_threads{0};
+
+				f32 ppu_usage{0};
+				f32 spu_usage{0};
+				f32 rsx_usage{0};
+				u32 rsx_load{0};
+
+				std::shared_ptr<GSRender> rsx_thread;
+
+				std::string perf_text;
+
+				// 1. Fetch/calculate metrics we'll need
+				switch (m_detail)
+				{
+				case detail_level::ultra_5:
+					rsx_thread = fxm::get<GSRender>();
+					rsx_load   = rsx_thread->get_load();
+
+					/* fallthrough */
+				case detail_level::high_4:
+					frametime     = force_update ? 0 : std::max(0.0, elapsed / m_frames);
+					total_threads = CPUStats::get_thread_count();
+
+					/* fallthrough */
+				case detail_level::medium_3:
+					ppus = idm::select<ppu_thread>([&ppu_cycles](u32, ppu_thread& ppu) { ppu_cycles += ppu.get()->get_cycles(); });
+
+					spus = idm::select<SPUThread>([&spu_cycles](u32, SPUThread& spu) { spu_cycles += spu.get()->get_cycles(); });
+
+					rawspus = idm::select<RawSPUThread>([&spu_cycles](u32, RawSPUThread& rawspu) { spu_cycles += rawspu.get()->get_cycles(); });
+
+					if (!rsx_thread)
+						rsx_thread = fxm::get<GSRender>();
```

Comment:
use fxm::get_always

Reply:
Actually never forcefully create rsx::thread which is the base for GSRender. Doesn't matter anyway since a better way will exist soon (via rsx::get_current_renderer()). If an instance of rsx does not exist (which is impossible since update is called from rsx->flip), just skip getting its cycles count since obviously its 0.

Reply:
The reason I check that here is because `rsx_thread` might be already present from the switch fallthrough above

The point of the fallthrough thing is to avoid code duplication, but it does cause a few things like this. Hopefully not too bad to the point of it being unmaintainable though.